File: frontend/src/types/metaAgent.ts
/**
 * Meta-Agent FUZZY TypeScript Types
 * 
 * Defines TypeScript interfaces matching backend schemas for the meta-agent API.
 */

// ============== Permission Levels ==============

export enum PermissionLevel {
    VIEWER = "viewer",
    EDITOR = "editor",
    ADMIN = "admin",
    SUPERUSER = "superuser"
}

// ============== Command Status ==============

export enum CommandStatus {
    PENDING = "pending",
    EXECUTING = "executing",
    COMPLETED = "completed",
    FAILED = "failed"
}

// ============== Session Status ==============

export enum SessionStatus {
    ACTIVE = "active",
    COMPLETED = "completed",
    TIMEOUT = "timeout"
}

// ============== MetaAgent Types ==============

export interface MetaAgent {
    id: number;
    name: string;
    description?: string;
    is_active: boolean;
    permission_level: PermissionLevel;
    allowed_actions?: string[];
    configuration?: Record<string, unknown>;
    created_at: string;
    updated_at: string;
}

export interface MetaAgentCreate {
    name: string;
    description?: string;
    is_active?: boolean;
    permission_level?: PermissionLevel;
    allowed_actions?: string[];
    configuration?: Record<string, unknown>;
}

export interface MetaAgentUpdate {
    name?: string;
    description?: string;
    is_active?: boolean;
    permission_level?: PermissionLevel;
    allowed_actions?: string[];
    configuration?: Record<string, unknown>;
}

// ============== MetaAgentCommand Types ==============

export interface MetaAgentCommand {
    id: number;
    meta_agent_id: number;
    session_id?: string;
    command_type: string;
    intent: string;
    parameters?: Record<string, unknown>;
    status: CommandStatus;
    result?: Record<string, unknown>;
    error_message?: string;
    execution_time_ms?: number;
    created_at: string;
}

export interface MetaAgentCommandCreate {
    meta_agent_id: number;
    session_id?: string;
    command_type: string;
    intent: string;
    parameters?: Record<string, unknown>;
}

// ============== MetaAgentSession Types ==============

export interface MetaAgentSession {
    id: string;
    user_id: number;
    meta_agent_id: number;
    status: SessionStatus;
    context?: SessionContext;
    created_at: string;
    updated_at: string;
    completed_at?: string;
}

export interface MetaAgentSessionCreate {
    meta_agent_id: number;
}

export interface SessionContext {
    history: CommandHistoryEntry[];
    metadata?: Record<string, unknown>;
}

export interface CommandHistoryEntry {
    command: string;
    intent: string;
    result: Record<string, unknown>;
    timestamp: string;
}

// ============== Command Execution Types ==============

export interface CommandExecutionRequest {
    command: string;
    parameters?: Record<string, unknown>;
    session_id?: string;
}

export interface CommandExecutionResponse {
    session_id: string;
    result: Record<string, unknown>;
    execution_time_ms: number;
    command_id?: number;
}

// ============== Command List Types ==============

export interface CommandListResponse {
    commands: MetaAgentCommand[];
    total: number;
    limit: number;
    offset: number;
}

// ============== Session History Response ==============

export interface SessionHistoryResponse {
    session: MetaAgentSession;
    history: MetaAgentCommand[];
    context: SessionContext;
}

// ============== Pagination Types ==============

export interface PaginationParams {
    limit?: number;
    offset?: number;
}

export interface CommandFilters extends PaginationParams {
    session_id?: string;
    status?: CommandStatus;
}

// ============== UI State Types ==============

export interface CommandExecutionState {
    isExecuting: boolean;
    currentCommand?: string;
    error?: string;
    executionTime?: number;
}

export interface SessionState {
    currentSession?: MetaAgentSession;
    isLoading: boolean;
    error?: string;
}

export interface CommandHistoryItem {
    id: number;
    command: string;
    intent: string;
    status: CommandStatus;
    executionTime: number;
    result?: Record<string, unknown>;
    error?: string;
    timestamp: string;
    expanded?: boolean;
}

File: backend/app/schemas/meta_agent.py
"""
Meta-Agent FUZZY Pydantic Schemas

Defines request/response schemas for meta-agent API endpoints.
"""
from pydantic import BaseModel, Field
from typing import Optional, List, Dict, Any
from datetime import datetime
from enum import Enum


class PermissionLevel(str, Enum):
    """Permission levels for meta-agent actions"""
    VIEWER = "viewer"
    EDITOR = "editor"
    ADMIN = "admin"
    SUPERUSER = "superuser"


class CommandStatus(str, Enum):
    """Status of meta-agent command execution"""
    PENDING = "pending"
    EXECUTING = "executing"
    COMPLETED = "completed"
    FAILED = "failed"


class SessionStatus(str, Enum):
    """Status of meta-agent session"""
    ACTIVE = "active"
    COMPLETED = "completed"
    TIMEOUT = "timeout"


# ============== MetaAgent Schemas ==============

class MetaAgentBase(BaseModel):
    """Base schema for MetaAgent"""
    name: str = Field(..., min_length=1, max_length=100)
    description: Optional[str] = None
    is_active: bool = True


class MetaAgentCreate(MetaAgentBase):
    """Schema for creating a new MetaAgent"""
    permission_level: PermissionLevel = PermissionLevel.EDITOR
    allowed_actions: Optional[List[str]] = None
    configuration: Optional[Dict[str, Any]] = None


class MetaAgentUpdate(BaseModel):
    """Schema for updating a MetaAgent"""
    name: Optional[str] = Field(None, min_length=1, max_length=100)
    description: Optional[str] = None
    is_active: Optional[bool] = None
    permission_level: Optional[PermissionLevel] = None
    allowed_actions: Optional[List[str]] = None
    configuration: Optional[Dict[str, Any]] = None


class MetaAgentResponse(MetaAgentBase):
    """Schema for MetaAgent response"""
    id: int
    permission_level: PermissionLevel
    allowed_actions: Optional[List[str]] = None
    configuration: Optional[Dict[str, Any]] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


# ============== MetaAgentCommand Schemas ==============

class MetaAgentCommandBase(BaseModel):
    """Base schema for MetaAgentCommand"""
    command_type: str
    intent: str
    parameters: Optional[Dict[str, Any]] = None


class MetaAgentCommandCreate(MetaAgentCommandBase):
    """Schema for creating a new MetaAgentCommand"""
    meta_agent_id: int
    session_id: Optional[str] = None


class MetaAgentCommandResponse(MetaAgentCommandBase):
    """Schema for MetaAgentCommand response"""
    id: int
    meta_agent_id: int
    session_id: Optional[str] = None
    status: CommandStatus
    result: Optional[Dict[str, Any]] = None
    error_message: Optional[str] = None
    execution_time_ms: Optional[float] = None
    created_at: datetime

    class Config:
        from_attributes = True


# ============== MetaAgentSession Schemas ==============

class MetaAgentSessionBase(BaseModel):
    """Base schema for MetaAgentSession"""
    pass


class MetaAgentSessionCreate(BaseModel):
    """Schema for creating a new MetaAgentSession"""
    meta_agent_id: int


class MetaAgentSessionResponse(BaseModel):
    """Schema for MetaAgentSession response"""
    id: int
    user_id: int
    meta_agent_id: int
    status: SessionStatus
    context: Optional[Dict[str, Any]] = None
    created_at: datetime
    updated_at: datetime
    completed_at: Optional[datetime] = None

    class Config:
        from_attributes = True


# ============== Command Execution Schemas ==============

class CommandExecutionRequest(BaseModel):
    """Schema for executing a meta-agent command"""
    command: str = Field(..., description="Natural language command to execute")
    parameters: Optional[Dict[str, Any]] = None
    session_id: Optional[str] = Field(None, description="Optional session ID for continuing a conversation")


class CommandExecutionResponse(BaseModel):
    """Schema for command execution response"""
    session_id: str
    result: Dict[str, Any]
    execution_time_ms: float
    command_id: Optional[int] = None


# ============== Pagination Schemas ==============

class CommandListResponse(BaseModel):
    """Schema for paginated command list"""
    commands: List[MetaAgentCommandResponse]
    total: int
    limit: int
    offset: int